### ISSUE TITLE
What's the softmax temperature?

### ISSUE THREAD

**Issue by AlbertoSinigaglia** [idiap/attention-sampling#23]
I'm building this from scratch to avoid the additional CPP code, and it seems to be working, however, when I compute   
<br>
![](https://latex.codecogs.com/png.latex?%5Cdpi%7B300%7D%20%5Csum%20a%28x%29f%28x%29)
<br>
and use that as logits for the softmax, the "a" term on large images tend to be very small, and "f" is normalized, so the scalar product gets logits very close to 0, and since softmax is not scale invariant, I get close to uniform predictions...

Thus, I think that you are using some form of temperature, but nor in the code neither in the paper i see any reference to it... can I have some clarification?

At the moment the best I was able to to do (not to handpick the temperature) is to make it trainable:
```python
class SoftmaxWithTemperature(tf.keras.layers.Layer):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.t = self.add_weight("temperature", (1,), initializer=tf.initializers.ones())
    def call(self, inputs):
        return tf.nn.softmax(self.t * inputs)
classification_network = tf.keras.models.Sequential([
    SoftmaxWithTemperature()
])
classification_network(tf.reshape(features * probabilities, ...))
```

And while training, I see the temperature slowly increasing to 1.2, which is still not enough (at least in my case)

**Comment by AlbertoSinigaglia** [comment]
I ended up normalizing the expected feature like this:
<br>
![](https://latex.codecogs.com/svg.image?\huge&space;&space;o_j&space;=&space;softmax\left(\frac{\sum_{i=0}^P&space;a(x;\theta)_{\text{patch}_i}&space;\cdot&space;f(\text{patch}_i;\phi)}{\sum_{i=0}^P&space;a(x;\theta)_{\text{patch}_i}}\right))

With this normalization, no temperature is needed at it seems to have an "expected good performance"
